### DISCUSSION TITLE
[WIP] Dive into Deep Learning: TorchSharp Edition

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#1301] [General]
### TorchSharp Edition
> Just found out that there is a very early stage to bring a popular Deep AI course used in etc. colleges and universities.
> Once it is more progress, I will share the link here. Thank you to the one who has the vision to start it.

- [动手学深度学习](https://github.com/d2l-ai/d2l-zh)
- [Dive into Deep Learning](https://github.com/d2l-ai/d2l-en)
![image](https://github.com/dotnet/TorchSharp/assets/49812372/3a57b70f-1bb5-455b-aafb-a2742dfa8565)

**Comment by yueyinqiu** [comment]
I am doing some relevant things [here](https://github.com/yueyinqiu/d2l-TorchSharp-zh).

My original intentions are to:
- showcase the ability of TorchSharp
- help finding some problems of TorchSharp, e.g. #1294
- force myself to learn deep learning (Sadly, this seems of no use. I'm just a porter of the codes.)

Actually there are too many difference between Python and C#, and I don't have much time to rewrite the whole book. So I'm just rewriting the codes and making some explains (when something has too be done differently from the original book) in Chinese.

If anyone wants to do more about this, I'm glad to help.

**Reply by GeorgeS2019** [reply]
We started with one of the ideas that one day TorchSharp will be used in teaching. 
Everyone come along contributes a little bit.
The confidence of this project grows to the extends that its existence becoming clear.

Many efforts here do not necessary complete to 100%.

Other users come along later to push it closer to 100%

There is no pressure on your part.

Just enjoy "Learning" and that is the key!

**Reply by GeorgeS2019** [reply]
> Sadly, this seems of no use. I'm just a porter of the codes.

NO!  I have seen you making significant contribution. I thank you for your effort.

**Reply by GeorgeS2019** [reply]
@yueyinqiu 

I look through one of the lectures using the syllabus.
https://courses.d2l.ai/berkeley-stat-157/syllabus.html

![image](https://github.com/dotnet/TorchSharp/assets/49812372/ba9acd10-47ae-4be4-bcb3-53ef56b645f3)

These are typical real-world use cases: 
- https://courses.d2l.ai/berkeley-stat-157/units/word2vec.html
- https://courses.d2l.ai/berkeley-stat-157/units/seq2seq.html
- https://courses.d2l.ai/berkeley-stat-157/units/adam.html

**Comment by GeorgeS2019** [comment]
> showcase the ability of TorchSharp

The biggest use case in my view is in AI for game. 
Currently, most game engines support c#
However to do AI, these game engines need to support communication between e.g. pytorch and c# codes

With TorchSharp, no more such communication needed

**Comment by yueyinqiu** [comment]
I just found that there are many difference between the Chinese version and English version... Even the number of chapters are not the same.

**Reply by GeorgeS2019** [reply]
OK, I will take a look. thx for pointing that out.

**Reply by GeorgeS2019** [reply]
Chinese
1st Edition (10 chapters) 2nd Edition (15 chapters)
1st Edition chapter 9 => 2nd Edition chapter 13  (Deep Computational Vision)
1st Edition chapter 10 (Deep NLP)  => 2nd Edition chapter 14 (Deep NLP) and the extra chapter15 ( Application of Deep NLP in real world)

Chinese 2nd Edition => English 1st Edition  [It seems the English 1st Edition follows the Chinese 2nd Edition in chapters BUT with additional 16 to 21 chapters) 
#### Chinese 2nd Edition https://zh.d2l.ai/
#### English 1st Edition: https://d2l.ai/
Chinese 2nd Edition chapter 13 => English chapter 14 (Deep Computational Vision)
Chinese 2nd Edition chapter 14 => English chapter 15 (Deep NLP) and the extra chapter16 ( [Natural Language Processing: Applications](https://d2l.ai/chapter_natural-language-processing-applications/index.html) in real world)

The English version has additional chapters
17
18
19
20
21

**Reply by GeorgeS2019** [reply]
@yueyinqiu 
You have recently helped to organize TorchSharp => chapter 14 (English Version)

There are many users here already contributed to chapters 15 and 16 (English version). However, these contributions have yet to be organized into TorchSharp version of => TorchText

**Reply by GeorgeS2019** [reply]
@yueyinqiu 
I just made new changes. It is better to follow the English version: more up-to-date to the latest Deep AI needs, with additional chapters (16 to 21)

**Reply by yueyinqiu** [reply]
@GeorgeS2019

Thanks for your checking. I have tried a bit. But it tured out to be a little hard for me to read/write such unfamiliar things in English. Since the difference are almost additions only, perhaps I shall first continue with the Chinese version, and then add new things according to the English one.
